File: src/main/java/com/projectswg/forwarder/resources/networking/NetInterceptor.java
package com.projectswg.forwarder.resources.networking;

import com.projectswg.common.data.encodables.galaxy.Galaxy;
import com.projectswg.common.network.NetBuffer;
import com.projectswg.common.network.packets.PacketType;
import com.projectswg.common.network.packets.swg.login.LoginClientId;
import com.projectswg.common.network.packets.swg.login.LoginClusterStatus;
import com.projectswg.forwarder.Forwarder.ForwarderData;
import me.joshlarson.jlcommon.log.Log;
import me.joshlarson.jlcommon.utilities.ByteUtilities;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

public class NetInterceptor {
	
	private final ForwarderData data;
	
	public NetInterceptor(ForwarderData data) {
		this.data = data;
	}
	
	public byte[] interceptClient(byte[] data) {
		if (data.length < 6)
			return data;
		ByteBuffer bb = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN);
		PacketType type = PacketType.fromCrc(bb.getInt(2));
		switch (type) {
			case LOGIN_CLIENT_ID:
				return setAutoLogin(NetBuffer.wrap(bb));
			default:
				return data;
		}
	}
	
	public byte[] interceptServer(byte[] data) {
		if (data.length < 6)
			return data;
		ByteBuffer bb = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN);
		PacketType type = PacketType.fromCrc(bb.getInt(2));
		switch (type) {
			case LOGIN_CLUSTER_STATUS:
				return getServerList(NetBuffer.wrap(bb));
			default:
				return data;
		}
	}
	
	private byte[] setAutoLogin(NetBuffer data) {
		LoginClientId id = new LoginClientId(data);
		if (!id.getUsername().equals(this.data.getUsername()) || !id.getPassword().isEmpty())
			return data.array();
		id.setPassword(this.data.getPassword());
		return id.encode().array();
	}
	
	private byte[] getServerList(NetBuffer data) {
		LoginClusterStatus cluster = new LoginClusterStatus();
		cluster.decode(data);
		for (Galaxy g : cluster.getGalaxies()) {
			g.setAddress("127.0.0.1");
			g.setZonePort(this.data.getZonePort());
			g.setPingPort(this.data.getZonePort());
		}
		return cluster.encode().array();
	}
	
}

File: src/main/kotlin/com/projectswg/forwarder/resources/networking/NetInterceptor.kt
package com.projectswg.forwarder.resources.networking

import com.projectswg.common.network.NetBuffer
import com.projectswg.common.network.packets.PacketType
import com.projectswg.common.network.packets.swg.holo.login.HoloLoginRequestPacket
import com.projectswg.common.network.packets.swg.login.LoginClientId
import com.projectswg.common.network.packets.swg.login.LoginClusterStatus
import com.projectswg.forwarder.Forwarder.ForwarderData
import com.projectswg.holocore.client.HolocoreSocket
import java.nio.ByteBuffer
import java.nio.ByteOrder

class NetInterceptor(private val data: ForwarderData) {
	
	fun interceptClient(holocore: HolocoreSocket, data: ByteArray) {
		if (data.size < 6) {
			holocore.send(data)
			return
		}
		val bb = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)
		
		when (PacketType.fromCrc(bb.getInt(2))) {
			PacketType.LOGIN_CLIENT_ID -> {
				val loginClientId = LoginClientId(NetBuffer.wrap(bb))
				if (loginClientId.username == this.data.username && loginClientId.password.isEmpty())
					loginClientId.password = this.data.password
				holocore.send(HoloLoginRequestPacket(loginClientId.username, loginClientId.password).encode().array())
			}
			else -> holocore.send(data)
		}
	}
	
	fun interceptServer(data: ByteArray): ByteArray {
		if (data.size < 6)
			return data
		val bb = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)
		when (PacketType.fromCrc(bb.getInt(2))) {
			PacketType.LOGIN_CLUSTER_STATUS -> return getServerList(NetBuffer.wrap(bb))
			else -> return data
		}
	}
	
	private fun getServerList(data: NetBuffer): ByteArray {
		val cluster = LoginClusterStatus()
		cluster.decode(data)
		for (g in cluster.galaxies) {
			g.address = "127.0.0.1"
			g.zonePort = this.data.zonePort
			g.pingPort = this.data.pingPort
		}
		return cluster.encode().array()
	}
	
}
